staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go

```diff
@@ -599,6 +607,34 @@ func (s preparedGenericAPIServer) Run(stopCh <-chan struct{}) error {
 		s.NonLongRunningRequestWaitGroup.Wait()
 	}()
 
+	// wait for all in-flight watches to finish
+	activeWatchesDrainedCh := make(chan struct{})
+	go func() {
+		defer close(activeWatchesDrainedCh)
+
+		<-notAcceptingNewRequestCh.Signaled()
+
+		// Wait for all active watches to finish
+		grace := 10 * time.Second
+		activeBefore, activeAfter, err := s.WatchRequestWaitGroup.Wait(func(count int) (utilwaitgroup.RateLimiter, context.Context, context.CancelFunc) {
+			ctx, cancel := context.WithTimeout(context.Background(), grace)
+			// TODO: a) do we need to make this grace period configurable?
+			//  b) we need to tweak the qps to be more robust.
+			qps := count / int(grace.Seconds())
+			return rate.NewLimiter(rate.Limit(qps), 100), ctx, cancel
+		})
+		klog.V(1).InfoS("[graceful-termination] active watch request(s) have drained",
+			"grace-period", grace, "active-watches(pre wait)", activeBefore, "active-watches(post wait)", activeAfter, "error", err)
+	}()
+
+	go func() {
+		defer klog.V(1).InfoS("[graceful-termination] shutdown event", "name", drainedCh.Name())
+		defer drainedCh.Signal()
+
+		<-nonLongRunningRequestDrainedCh
+		<-activeWatchesDrainedCh
```

Comment:
We need to update the diagram in lines 436-470

Reply:
updated the diagram